pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,305 @@
+		return err
+	}
+	if len(oldTsDir) > 0 {
+		err = os.RemoveAll(path.Join(w.TargetDir, oldTsDir))
+		if err != nil {
+			glog.Errorf("%v: unable to remove directory %v: %v", w.LogContext, oldTsDir, err)
+		}
+	}
+
+	// TODO: determine error handling around sentinel file
+	return w.touchSentinelFile()
+}
+
+// shouldWritePayload returns whether the payload must be written to disk
+func (w *AtomicWriter) shouldWritePayload(payload map[string][]byte) (bool, error) {
+	for projectionPath, content := range payload {
+		shouldWrite, err := w.shouldWriteFile(path.Join(w.TargetDir, path.Clean(projectionPath)), content)
+		if err != nil {
+			return false, err
+		}
+
+		if shouldWrite {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// shouldWriteFile returns whether a new version of a file should be written to disk.
+func (w *AtomicWriter) shouldWriteFile(path string, content []byte) (bool, error) {
+	_, err := os.Lstat(path)
+	if err != nil && os.IsNotExist(err) {
+		return true, nil
+	}
+
+	contentOnFs, err := ioutil.ReadFile(path)
+	if err != nil {
+		return false, err
+	}
+
+	return (bytes.Compare(content, contentOnFs) != 0), nil
+}
+
+// newTimestampDir creates a new timestamped directory
+func (w *AtomicWriter) newTimestampDir() (string, error) {
+	tsDir, err := ioutil.TempDir(w.TargetDir, fmt.Sprintf("..%v", ts()))
+	if err != nil {
+		glog.Errorf("%v: unable to create new temp directory: %v", w.LogContext, err)
+		return "", err
+	}
+
+	return tsDir, nil
+}
+
+// writePayloadToDir writes the given payload to the given directory.  The
+// directory must exist.
+func (w *AtomicWriter) writePayloadToDir(payload map[string][]byte, dir string) error {
+	allErrs := []error{}
+	for projectionPath, content := range payload {
+		fullPath := path.Join(dir, path.Clean(projectionPath))
+		dir, _ := filepath.Split(fullPath)
+
+		err := os.MkdirAll(dir, os.ModePerm)
+		if err != nil {
+			glog.Errorf("%v: unable to create directory %v: %v", w.LogContext, dir, err)
+			allErrs = append(allErrs, err)
+		}
+
+		err = ioutil.WriteFile(fullPath, content, 0644)
+		if err != nil {
+			glog.Errorf("%v: unable to write file %v: %v", w.LogContext, fullPath, err)
+			allErrs = append(allErrs, err)
+		}
+	}
+
+	return utilerrors.NewAggregate(allErrs)
+}
+
+// createPayloadSymlinksInTargetDir creates the relative symlinks for all the
+// files configured in the payload. If the directory in a file path does not
+// exist, it is created.
+//
+// Viz:
+// For files: "bar", "foo/bar", "baz/bar", "foo/baz/blah"
+// the following symlinks and subdirectories are created:
+// bar          -> ..identifierToken/bar
+// foo/bar      -> ../..identifierToken/foo/bar
+// baz/bar      -> ../..identifierToken/baz/bar
+// foo/baz/blah -> ../../..identifierToken/foo/baz/blah
+func (w *AtomicWriter) createPayloadSymlinksInTargetDir(payload map[string][]byte) error {
+	for projectionPath := range payload {
+		cleanProjectionPath := path.Clean(projectionPath)
```

Comment:
you should just Clean paths when you construct the object - sanitize it once into a private field.